src/middlewared/middlewared/plugins/init_shutdown_script.py

```diff
@@ -156,72 +143,71 @@ async def validate(self, data, schema_name):
 
         verrors.check()
 
+    @private
+    def write_script_text(self, script_text):
+        with tempfile.NamedTemporaryFile(delete=False, mode='w') as f:
+            f.write(script_text)
+            os.fchmod(f.fileno(), stat.S_IRWXU)
+            return f.name
+
+    @private
+    def script_already_exists(self, script):
+        return os.path.exists(script) and os.access(script, os.X_OK)
+
+    @private
+    def rem_tmp_script(self, script):
+        try:
+            os.unlink(script)
+        except FileNotFoundError:
+            pass
+        except Exception:
+            self.logger.debug('Unexpected failure removing %r', script, exc_info=True)
+
     @private
     async def execute_task(self, task):
         task_type = task['type']
-        cmd = None
-        tmp_script = None
+        cmd = tmp_script = None
 
         if task_type == 'COMMAND':
             cmd = task['command']
         elif task_type == 'SCRIPT' and task['script_text']:
-            fd, tmp_script = tempfile.mkstemp(text=True)
-            os.close(fd)
-            os.chmod(tmp_script, stat.S_IRWXU)
-            with open(tmp_script, 'w') as f:
-                f.write(task['script_text'])
+            tmp_script = await self.middleware.run_in_thread(self.write_script_text, task['script_text'])
             cmd = f'exec {tmp_script}'
-        elif os.path.exists(task['script'] or '') and os.access(task['script'], os.X_OK):
+        elif await self.middleware.run_in_thread(self.script_already_exists, task['script']):
```

Comment:
it exists??

Reply:
Why `script_already_exists` and not just `script_exists`? `already` makes reader think we do this check to avoid creating something that already exists which is not the case here.